File: src/server/gaia_v2/util/hashing.ts
import type { ReadonlyVec3 } from "@/shared/math/types";

function randomHash(x: number) {
  x ^= x >> 16;
  x *= 0x7feb352d;
  x ^= x >> 15;
  x *= 0x846ca68b;
  x ^= x >> 16;
  return x;
}

// Same as the version in C++
export function positionHash([x, y, z]: ReadonlyVec3) {
  return randomHash(x + randomHash(y + randomHash(z)));
}

File: src/galois/py/assets/impl/hashing.py
import numpy as np


def random_hash(x: int):
    x = int(x)
    x ^= x >> 16
    x *= 0x7FEB352D
    x ^= x >> 15
    x *= 0x846CA68B
    x ^= x >> 16
    return x


def position_hash(x: int, y: int, z: int):
    return random_hash(x + random_hash(y + random_hash(z)))
